tools/cli/lib/api.c

```diff
@@ -671,6 +671,211 @@ TDNFCliRepoSyncCommand(
+        pszVal = pPkgInfo->pszSourcePkg;
+    } else if (strcmp(tag, "description") == 0) {
+        pszVal = pPkgInfo->pszDescription;
+    } else if (strcmp(tag, "summary") == 0) {
+        pszVal = pPkgInfo->pszSummary;
+    } else if (strcmp(tag, "license") == 0) {
+        pszVal = pPkgInfo->pszLicense;
+    } else if (strcmp(tag, "url") == 0) {
+        pszVal = pPkgInfo->pszURL;
+    } else if (strcmp(tag, "provides") == 0) {
+        dwError = parse_depResult(pPkgInfo->ppszDependencies[0], &pszVal);
+    } else if (strcmp(tag, "obsoletes") == 0) {
+        dwError = parse_depResult(pPkgInfo->ppszDependencies[1], &pszVal);
+    } else if (strcmp(tag, "conflicts") == 0) {
+        dwError = parse_depResult(pPkgInfo->ppszDependencies[2], &pszVal);
+    } else if (strcmp(tag, "requires") == 0) {
+        dwError = parse_depResult(pPkgInfo->ppszDependencies[3], &pszVal);
+    } else {
+        dwError = ERROR_TDNF_INVALID_PARAMETER;
+        BAIL_ON_TDNF_ERROR(dwError);
+    }
+
+    if (pszVal) {
+        *pszRet = strdup(pszVal);
+    } else {
+        *pszRet = strdup("(null)");
+    }
+
+cleanup:
+    return dwError;
+error:
+    goto cleanup;
+}
+
+/* similar semantics as snprintf() */
+uint32_t format_string(char **out, size_t *len, const char *format,
+                  uint32_t (*tag_handler)(const char *, void *, char **), void *data)
+{
+    const char *p;
+    char *q;
+    int l = 0;
+    int nLength = 0;
+    uint32_t dwError = 0;
+    int buffer_bytes = 20;
+
+    p = format;
+    q = *out;
+    while (*p) {
+        if (*p == '%' && p[1] == '{') {
+            const char *t = &p[2];
+            char *tag, *value;
+            size_t tag_len;
+
+            while(*t && *t != '}') t++;
+
+            if (!*t)
+                return -1;
+            tag_len = t - &p[2];
+            tag = (char *)malloc(tag_len+1);
+            strncpy(tag, &p[2], tag_len);
+            tag[tag_len] = 0;
+            value = NULL;
+ 
+            dwError = tag_handler(tag, data, &value);
+            if (dwError == ERROR_TDNF_INVALID_PARAMETER) {
+                pr_err("Package doesn't have attribute %s\n", tag);
+            }
+            BAIL_ON_TDNF_ERROR(dwError);
+
+            if ((*len <= strlen(value))  || ((*len >= strlen(*out)) && (strlen(value) > (*len - strlen(*out))))) {
+                nLength = strlen(*out) + strlen(value) + buffer_bytes;
+                dwError = TDNFReAllocateMemory(nLength, (void **)out);
+                BAIL_ON_TDNF_ERROR(dwError);
+                *len = nLength;
+                q = *out;
+                q = q + l;
+            }
+
+            if (value) {
+                t = value;
+                while(*t) {
+                    *q = *t++; l++;
+                    if (q < *out + *len-1) {
+                        q++; *q = 0;
+                    }
+                }
+                *q = 0;
+                free(value);
+            }
+            p += tag_len + 3;
+
+        } else {
```

Comment:
We should handle C format like special characters like `\n` and so. Maybe you can copy code from `jsondump/jsondump.c `, in `jsonify_string()`.

Reply:
Sorry for pointing at `jsondump/jsondump.c` -it's the other way around.

If the format string contains "\n", it should be converted to a line feed. Similar for the others. So you need to scan the format string for backwards slashes (`if (*p=='\\')`), then see if the following char is a literal 'n'. If so, convert to a LF, like `'\n'`.

`dnf` (Fedora) does it correctly:
```
# dnf repoquery --qf "start\n%{url}\nend" dnf
Last metadata expiration check: 0:05:58 ago on Mon May 15 18:39:33 2023.
start
https://github.com/rpm-software-management/dnf
end
```

You implementation does not:
```
# ./bin/tdnf repoquery --qf "start\n%{url}\nend" tdnf
start\nhttps://github.com/vmware/tdnd
```
And it also just drops the "end". Something else is not right there.

Reply:
Changed the code and Verified special characters are handled properly in the output result.

Reply:
Thank you!